Added support for Common Function Attributes

shubhamnarlawar77 opened the pull request.

shubhamnarlawar77 committed:
Added support for Common Function Attributes

shubhamnarlawar77 commented:
@marxin : review request

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,57 @@
+}
+
+void
+Attribute::OutputAttribute(std::ostream &out, string option, AttributeGenerator &attr_gen)
```
I don't like that you pass `AttributeGenerator` here. `Attribute` class should not know about `AttributeGenerator`.

shubhamnarlawar77 replied:
Done. Passed bool variable `attr_emitted` instead of object.

marxin replied:
No, `OutputAttribute` should not be passed such an argument. Let the Generator print `"__attribute__ (("` and corresponding ending.

shubhamnarlawar77 replied:
But generator does not know how many attributes will pass `rnd_flipcoin()` probability. It can only print individual attributes like `__attribute__((hot)) __attribute__((cold))` in current state if we dont pass `object` or `attr_emitted`.

marxin replied:
What about `OutputAttribute` function indicating whether it adds an attribute or not?

shubhamnarlawar77 replied:
Like returning a bool variable from `OutputAttribute` whether an attribute have passed `rnd_flipcoin()` in `OutputAttributes()`?

marxin replied:
Well, I would prefer `OutputAttribute` returns a string which can then be printed to the stream by `AttributeGenerator`.

shubhamnarlawar77 replied:
Done. Please check it.

marxin commented on src/Function.cpp:
```diff
@@ -562,6 +583,13 @@ Function::OutputForwardDecl(std::ostream &out)
 	if(func_attr_inline)
 		out << " __attribute__((always_inline))";
+	vector<Attribute*>::iterator itr;
+	for(itr = func_attr_generator.attributes.begin(); itr < func_attr_generator.attributes.end(); itr++)
```
Call something like `func_attr_generator.GenerateAllAttributes()`/

shubhamnarlawar77 replied:
> Call something like `func_attr_generator.GenerateAllAttributes()`/

Got it.

marxin commented on src/Makefile.am:
```diff
@@ -60,6 +60,8 @@ csmith_SOURCES = \
 	AbsRndNumGenerator.h \
 	ArrayVariable.cpp \
 	ArrayVariable.h \
+	Attribute.cpp \
```
You probably forgot to re-generate `Makefile.in` as I advised you.

shubhamnarlawar77 replied:
I have added new file in `CMakeLists.txt` and `Makefile.am`. You said `you'all need to run automake to generate Makefile.in`. I didn't get this part. Can you elaborate a bit on how to do it?

marxin replied:
`marxin@marxinbox:~/Programming/csmith> automake`

shubhamnarlawar77 replied:
Done. Should I push it as a separate commit `Updated Makefile` in this pull request?

marxin replied:
Please include it in this pull request.

shubhamnarlawar77 replied:
Done. Please check it.

marxin replied:
Looks you have an old version of the `automake`, please use this instead: https://pastebin.com/LCXwXzaR and do not make it a new commit, include it where you added `Attribute.cpp` file.

shubhamnarlawar77 replied:
Thanks. Done.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,74 @@
+	vector<Attribute*>::iterator itr;
+	for(itr = attributes.begin(); itr < attributes.end(); itr++){
+		attr_generated = (*itr)->OutputAttributes();
+		if(!attr_emitted && attr_generated != ""){
```
Why do you compare here with `""`?

shubhamnarlawar77 replied:
To identify whether an attribute has passed `rnd_flipcoin()`. If it did not pass the string will be "".
Sure. I will make string `attr_generated` local.

marxin replied:
Ah sorry, the empty string comparison is fine.

marxin replied:
Make condition for `attr_generated != ""` first and later test for `attr_emitted`. Make `attr_emitted` a local variable in this if block.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,74 @@
+{
+	vector<Attribute*>::iterator itr;
+	for(itr = attributes.begin(); itr < attributes.end(); itr++){
+		attr_generated = (*itr)->OutputAttributes();
```
Can you please make it a local variable and not a class member variable?

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,74 @@
+	}
+	if(attr_emitted){
+		out << "))";
+		attr_emitted = false;
```
Why do you reset it?

shubhamnarlawar77 replied:
For next function , it must print from start i.e.  `__attribute__((`. Hence, resetting it.

marxin replied:
Then make it a local variable.

shubhamnarlawar77 replied:
Done. Please check it.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,74 @@
+}
+
+AttributeGenerator::AttributeGenerator()
+	: attr_emitted(false), attr_generated("")
```
Why do you use member variable `attr_generated`?

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,70 @@
+	bool attr_emitted = false;
+	string attr_generated = "";
+	vector<Attribute*>::iterator itr;
+	for(itr = attributes.begin(); itr < attributes.end(); itr++){
```
For iterators, you should compare for `!=` and not `<`.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,70 @@
+	}
+	if(attr_emitted){
+		out << "))";
+		attr_emitted = false;
```
Do you need this line?

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,70 @@
+BooleanAttribute::OutputAttributes()
+{
+	if(rnd_flipcoin(attribute_probability))
+		attr_generate = OutputAttribute("");
```
This is weird, can you please remove one of `OutputAttribute`, `OutputAttributes` functions?

shubhamnarlawar77 replied:
Removed `OutputAttribute`.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,70 @@
+string
+MultiValuedAttribute::OutputAttributes()
+{
+	if(rnd_flipcoin(attribute_probability))
```
Please rewrite it to `if (...) return something; else return "";`?

marxin commented on src/Attribute.h:
```diff
@@ -0,0 +1,43 @@
+	string attribute;
+	int attribute_probability;
+	string attr_generate;
+	Attribute(string, int);
```
Please declare also a parameter names, it will be more readible.

shubhamnarlawar77 replied:
Done.

shubhamnarlawar77 commented:
1. Changed iterator condition from < to !=
2. Checked attr_generated != "" and then later if-else for `attr_emitted`. But I can not move it to if block as I required it later to print `"))"` after for loop.
3. Removed attr_emitted = false line
4. Removed `OutputAttribute` from base class as only `OutputAttributes` is enough
5. Rewrited MultiValued function as `if (...) return something; else return "";`

Please check it.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,66 @@
+AttributeGenerator::GenerateAllAttributes(std::ostream &out)
+{
+	bool attr_emitted = false;
+	string attr_generated = "";
```
You don't need initialization and please put declaration into the loop body.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,66 @@
+{
+	if(rnd_flipcoin(attribute_probability))
+		attr_generate = attribute;
+	return attr_generate;
```
That's still bad.

marxin commented on src/Attribute.cpp:
```diff
@@ -0,0 +1,66 @@
+MultiValuedAttribute::OutputAttributes()
+{
+	if(rnd_flipcoin(attribute_probability)){
+		attr_generate = attribute + "(\"" + attribute_values[rnd_upto(attribute_values.size())] + "\")";
```
What about return `attribute + "(\"" + attribute_values[rnd_upto(attribute_values.size())] + "\")";` ?

marxin commented on src/Attribute.h:
```diff
@@ -0,0 +1,42 @@
+public:
+	string attribute;
+	int attribute_probability;
+	string attr_generate;
```
Remove this.

marxin commented on src/Attribute.h:
```diff
@@ -0,0 +1,42 @@
+
+class AttributeGenerator;
+
+class Attribute
```
Please write comments for member variables and functions. Likewise for class. I can be brief, but it would be useful for reviewers.

shubhamnarlawar77 replied:
Done. Please check updated changes.

marxin reviewed (changes requested):
Not it's much better, please fix suggested changes.

marxin commented on src/Attribute.h:
```diff
@@ -0,0 +1,41 @@
+	string attribute;				//Name of attribute
+	int attribute_probability;			//Attribute generation probability
+	Attribute(string, int);
+	virtual string OutputAttributes() = 0;		//Checks attribute probability and generate is accordingly
```
Please put the comment on a different line so that we don't have such long lines.
That's last change I see. Thanks.

shubhamnarlawar77 replied:
Done. I have placed those lines above respective classes with detailed description for readability. Is it fine now? 

Is pull request ready to merge?

marxin replied:
> Done. I have placed those lines above respective classes with detailed description for readability. Is it fine now?

No, please put the comments within the class declaration, just above a declaration.

shubhamnarlawar77 replied:
Done. Is it fine to merge now?

marxin replied:
Yes, it's fine. Thank you.
